### PR TITLE
Add readme for fabric-lifecycle-events-v1

### PR DESCRIPTION
**Pull request by i509VCB**
This adds a readme to the module which details all the events the module contains + some details about the events.

### CODE REVIEW

File: fabric-lifecycle-events-v1/README.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
+# Fabric Lifecycle Events (V1)
+
+## Organization
+
+Events in this module belong to a single class which relates to the category of events.
+The implementation of event callbacks are sub classes.
```
Comment (Juuxel) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
The implementation of event callbacks are nested classes.
```

**Review by Juuxel — CHANGES_REQUESTED** [review]

**Comment by OroArmor** [comment]
I like it, but I feel like the dropdown is just a little confusing and could be done differently. I see why it's done this way, but fully expanded, or something like a table, might be easier to find what you need.

**Comment by i509VCB** [comment]
I definitely want to see a nicer way to organize all of the module in the readme, however other methods feel cluttered. Though the dropdown can feel cluttered after opening up one or two sections

File: fabric-lifecycle-events-v1/README.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+- If a mod stores the server instance anywhere, this event should be used for reference cleanup.
+- Depending on the implementation of the server the following may occur after this event is called:
+    - On an integrated server (`EnvType.CLIENT`), the client will continue running.
+    - On a dedicated server (`EnvType.SERVER`), this will be the last event called. The JVM will terminate after this event.
+
+#### `ServerLifecycleEvents.START_DATA_PACK_RELOAD`
+- Called before a Minecraft server reloads data packs.
+- In vanilla, this is typically called after `/reload` command is executed.
+
+#### `ServerLifecycleEvents.END_DATA_PACK_RELOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerLifecycleEvents.END_DATA_PACK_RELOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+#### `ServerLifecycleEvents.SERVER_STOPPED`
+
+- Called when a Minecraft server has stopped.
+All worlds have been closed and all (block) entities and players have been unloaded.
+- If a mod stores the server instance anywhere, this event should be used for reference cleanup.
+- Depending on the implementation of the server the following may occur after this event is called:
+    - On an integrated server (`EnvType.CLIENT`), the client will continue running.
+    - On a dedicated server (`EnvType.SERVER`), this will be the last event called. The JVM will terminate after this event.
+
+#### `ServerLifecycleEvents.START_DATA_PACK_RELOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerLifecycleEvents.START_DATA_PACK_RELOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+- Called before a Minecraft server reloads data packs.
+- In vanilla, this is typically called after `/reload` command is executed.
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
- In vanilla, this is typically called after the `/reload` command is executed.
```

File: fabric-lifecycle-events-v1/README.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+#### `ServerLifecycleEvents.END_DATA_PACK_RELOAD`
+- Called after a Minecraft server has reloaded data packs.
+- This event does not garuntee the data pack reload was successful.
+- Per the `EndDataPackReload` interface, the last `boolean` parameter specifies whether the data pack reload was successful.
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]: Is `boolean` the actual parameter name? If not, it should either not be formatted or use the actual parameter name.

File: fabric-lifecycle-events-v1/README.md [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+```java
+@FunctionalInterface
+public interface EndDataPackReload {
+    void endDataPackReload(MinecraftServer server, ServerResourceManager serverResourceManager, boolean success);
+}
+```
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
````suggestion
```

````

File: fabric-lifecycle-events-v1/README.md [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+- If the data pack reload is successful, data packs will be set. If the reload failed, then the currently loaded data packs will be kept.
+</details>
+
+## `ServerTickEvents`
+
+<details><summary>Contains events related to the ticking of a Minecraft server.</summary>
+
+#### `ServerTickEvents.START_SERVER_TICK`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerTickEvents.START_SERVER_TICK`

```

File: fabric-lifecycle-events-v1/README.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+- Called at the start of the server tick.
+- This event can be used by mods to pre-process before the server tick occurs.
+
+#### `ServerTickEvents.END_SERVER_TICK`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerTickEvents.END_SERVER_TICK`

```

File: fabric-lifecycle-events-v1/README.md [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+- Called at the end of the server tick.
+
+#### `ServerTickEvents.START_WORLD_TICK`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerTickEvents.START_WORLD_TICK`

```

File: fabric-lifecycle-events-v1/README.md [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+- Called at the start of a ServerWorld's tick.
+
+#### `ServerTickEvents.END_WORLD_TICK`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerTickEvents.END_WORLD_TICK`

```

File: fabric-lifecycle-events-v1/README.md [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+- Called at the end of a ServerWorld's tick.
+- End of world tick may be used to start async computations for the next tick.
+</details>
+
+## `ServerWorldEvents`
+
+<details><summary>Events related to the lifecycle of server worlds.</summary>
+
+#### `ServerWorldEvents.LOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerWorldEvents.LOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+- Called when a world is loaded by a Minecraft server.
+- This event may be called at any time, but is typically called between `ServerLifecycleEvents.SERVER_STARTING` and `ServerLifecycleEvents.SERVER_STARTED`.
+- Mods which implement dynamically loaded dimensions may call this event to notify other mods of a new server world being loaded.
+
+#### `ServerWorldEvents.UNLOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerWorldEvents.UNLOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+- Called before a world is unloaded by a Minecraft server.
+- This event may be called at any time but is typically called after a server has started shutting down (`ServerLifecycleEvents.SERVER_STOPPING`).
+- Mods which implement dynamically loaded dimensions may call this event to notify other mods of a server world being unloaded for reference cleanup.
+</details>
+
+## `ServerChunkEvents`
+
+<details><summary>Events related to the lifecycle of chunks in a server world.</summary>
+
+#### `ServerChunkEvents.CHUNK_LOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerChunkEvents.CHUNK_LOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+#### `ServerChunkEvents.CHUNK_UNLOAD`
+- Called when an chunk is unloaded from a ServerWorld.
+- The chunk may be modified when this event is called.
+</details>
+
+## `ServerEntityEvents`
+
+<details><summary>Events related to the lifecycle of entities in a server world.</summary>
+
+#### `ServerEntityEvents.ENTITY_LOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerEntityEvents.ENTITY_LOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+
+## `ServerChunkEvents`
+
+<details><summary>Events related to the lifecycle of chunks in a server world.</summary>
+
+#### `ServerChunkEvents.CHUNK_LOAD`
+- Called when an chunk is loaded into a ServerWorld.
+- The chunk may be modified when this event is called.
+
+#### `ServerChunkEvents.CHUNK_UNLOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerChunkEvents.CHUNK_UNLOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+
+## `ClientBlockEntityEvents`
+
+<details><summary>Events related to the lifecycle of block entities in a client world</summary>
+
+#### `ClientBlockEntityEvents.BLOCK_ENTITY_LOAD`
+- Called when a BlockEntity is loaded into a client world.
+- The block entity is present in the client world when this event is called.
+
+#### `ClientBlockEntityEvents.BLOCK_ENTITY_LOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ClientBlockEntityEvents.BLOCK_ENTITY_LOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+#### `ClientEntityEvents.ENTITY_UNLOAD`
+- Called when an Entity is about to be unloaded from a client world.
+- The entity is present in the client world when this event is called.
+</details>
+
+## `ClientBlockEntityEvents`
+
+<details><summary>Events related to the lifecycle of block entities in a client world</summary>
+
+#### `ClientBlockEntityEvents.BLOCK_ENTITY_LOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ClientBlockEntityEvents.BLOCK_ENTITY_LOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+
+## `ClientEntityEvents`
+
+<details><summary>Events related to the lifecycle of entities in a client world</summary>
+
+#### `ClientEntityEvents.ENTITY_LOAD`
+- Called when an Entity is loaded into a client world.
+- The entity is present in the client world when this event is called.
+
+#### `ClientEntityEvents.ENTITY_UNLOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ClientEntityEvents.ENTITY_UNLOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+
+#### `ClientChunkEvents.CHUNK_UNLOAD`
+- Called when a chunk is about to be unloaded from a client world.
+</details>
+
+## `ClientEntityEvents`
+
+<details><summary>Events related to the lifecycle of entities in a client world</summary>
+
+#### `ClientEntityEvents.ENTITY_LOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ClientEntityEvents.ENTITY_LOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+</details>
+
+## `ClientChunkEvents`
+
+<details><summary>Events related to the lifecycle of chunks on a Minecraft client</summary>
+
+#### `ClientChunkEvents.CHUNK_LOAD`
+- Called when a chunk is loaded into a client world.
+
+#### `ClientChunkEvents.CHUNK_UNLOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ClientChunkEvents.CHUNK_UNLOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+#### `ClientTickEvents.END_WORLD_TICK`
+- Called at the end of a client world's tick.
+- End of world tick may be used to start async computations for the next tick.
+</details>
+
+## `ClientChunkEvents`
+
+<details><summary>Events related to the lifecycle of chunks on a Minecraft client</summary>
+
+#### `ClientChunkEvents.CHUNK_LOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ClientChunkEvents.CHUNK_LOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+#### `ClientTickEvents.START_CLIENT_TICK`
+- Called at the start of the client tick.
+
+#### `ClientTickEvents.END_CLIENT_TICK`
+- Called at the end of the client tick.
+
+#### `ClientTickEvents.START_WORLD_TICK`
+- Called at the start of a client world's tick.
+
+#### `ClientTickEvents.END_WORLD_TICK`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ClientTickEvents.END_WORLD_TICK`

```

File: fabric-lifecycle-events-v1/README.md [hunk 23]
```diff
@@ -0,0 +1,206 @@
...
+
+<details><summary>Events related to ticking of a Minecraft client.</summary>
+
+#### `ClientTickEvents.START_CLIENT_TICK`
+- Called at the start of the client tick.
+
+#### `ClientTickEvents.END_CLIENT_TICK`
+- Called at the end of the client tick.
+
+#### `ClientTickEvents.START_WORLD_TICK`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ClientTickEvents.START_WORLD_TICK`

```

File: fabric-lifecycle-events-v1/README.md [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+</details>
+
+## `ClientTickEvents`
+
+<details><summary>Events related to ticking of a Minecraft client.</summary>
+
+#### `ClientTickEvents.START_CLIENT_TICK`
+- Called at the start of the client tick.
+
+#### `ClientTickEvents.END_CLIENT_TICK`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ClientTickEvents.END_CLIENT_TICK`

```

File: fabric-lifecycle-events-v1/README.md [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+- Called when Minecraft's client begins to stop.
+- This is caused by quitting while in game, or closing the game window.
+- This is called before the integrated server is stopped.
+</details>
+
+## `ClientTickEvents`
+
+<details><summary>Events related to ticking of a Minecraft client.</summary>
+
+#### `ClientTickEvents.START_CLIENT_TICK`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ClientTickEvents.START_CLIENT_TICK`

```

File: fabric-lifecycle-events-v1/README.md [hunk 26]
```diff
@@ -0,0 +1,206 @@
...
+
+## `ClientLifecycleEvents`
+
+<details><summary>Events related to the lifecycle of a Minecraft client</summary>
+
+#### `ClientLifecycleEvents.CLIENT_STARTED`
+- Called when Minecraft has started and it's client about to tick for the first time.
+- This occurs while the splash screen is displayed.
+
+#### `ClientLifecycleEvents.CLIENT_STOPPING`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ClientLifecycleEvents.CLIENT_STOPPING`

```

File: fabric-lifecycle-events-v1/README.md [hunk 27]
```diff
@@ -0,0 +1,206 @@
...
+
+<details><summary>Events related to objects on a logical client.</summary>
+
+**Note: These events are only available on a client. Trying to access these events on a dedicated server will cause the game to crash.**
+
+## `ClientLifecycleEvents`
+
+<details><summary>Events related to the lifecycle of a Minecraft client</summary>
+
+#### `ClientLifecycleEvents.CLIENT_STARTED`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ClientLifecycleEvents.CLIENT_STARTED`

```

File: fabric-lifecycle-events-v1/README.md [hunk 28]
```diff
@@ -0,0 +1,206 @@
...
+
+## `ServerBlockEntityEvents`
+
+<details><summary>Events related to the lifecycle of block entities in a server world.</summary>
+
+#### `ServerBlockEntityEvents.BLOCK_ENTITY_LOAD`
+- Called when an BlockEntity is loaded into a ServerWorld.
+- The block entity is present in the server world when this event is called.
+
+#### `ServerBlockEntityEvents.BLOCK_ENTITY_UNLOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerBlockEntityEvents.BLOCK_ENTITY_UNLOAD`

```

File: fabric-lifecycle-events-v1/README.md [hunk 29]
```diff
@@ -0,0 +1,206 @@
...
+- Called when an Entity is loaded into a ServerWorld.
+- The entity is present in the server world when this event is called.
+- **The unload event has not been implemented yet.**
+</details>
+
+## `ServerBlockEntityEvents`
+
+<details><summary>Events related to the lifecycle of block entities in a server world.</summary>
+
+#### `ServerBlockEntityEvents.BLOCK_ENTITY_LOAD`
```
Comment (haykam821) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
#### `ServerBlockEntityEvents.BLOCK_ENTITY_LOAD`

```

**Review by haykam821 — CHANGES_REQUESTED** [review]
The mix of Markdown and HTML is unconventional so I'd recommend stripping away the `<details>` and `<summary>` tags, even though it's still valid.

It's better to have two newlines between a header and content as well. In addition, it seems like the headers skip from level 2 to level 4.

There is also a slight inconsistency with natural language ('server world') versus class names (`ServerWorld`). Perhaps we should standardize on one or the other, or even one only at the start of a section?

**Comment by Technici4n** [comment]
I think there should be a summary at the top of the README, or maybe on a wiki page listing the various modules? It would be nice to know roughly what each module does without needing to open source files I think.

**Comment by i509VCB** [comment]
I decided to take a different approach for this module. Since this module is rather wide in scope, I have decided to only document the categories, class names and then write a short sentence explaining what each of the classes contains. I believe the further depth could probably be on wiki or the javadoc should be consulted.

File: fabric-lifecycle-events-v1/README.md [hunk 30]
```diff
@@ -0,0 +1,78 @@
+# Fabric Lifecycle Events (V1)
+
+Contains events which indicate the when the lifecycle of a Minecraft Client, Minecraft Server and objects which exist on the client and server have changed.
```
Comment (Juuxel) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
Contains events which indicate the when the lifecycle of a Minecraft client, Minecraft server and objects which exist on the client and server have changed.
```

File: fabric-lifecycle-events-v1/README.md [hunk 31]
```diff
@@ -0,0 +1,78 @@
...
+For example, if you are listening to `ServerLifecycleEvents.SERVER_STARTING` in `ServerLifecycleEvents` the callback interface that would be implemented is `ServerLifecycleEvents.ServerStarting`.
+
+# Server Events
+
+Events related to objects on a logical server.
+These classes are located in `net.fabricmc.fabric.api.event.lifecycle.v1`.
+
+## `ServerLifecycleEvents`
+
+Contains lifecycle events related to a Minecraft Server.
```
Comment (Juuxel) [inline fabric-lifecycle-events-v1/README.md]:
```suggestion
Contains lifecycle events related to a Minecraft server.
```

**Review by Juuxel — CHANGES_REQUESTED** [review]

File: fabric-lifecycle-events-v1/README.md [hunk 32]
```diff
@@ -0,0 +1,78 @@
+# Fabric Lifecycle Events (V1)
+
+Contains events which indicate the when the lifecycle of a Minecraft client, Minecraft server and objects which exist on the client and server have changed.
```
Comment (Technici4n) [inline fabric-lifecycle-events-v1/README.md]: the when the?

Comment (Technici4n) [reply fabric-lifecycle-events-v1/README.md]: either the lifecycle has changed, or the lifecycles have changed (I think)

File: fabric-lifecycle-events-v1/README.md [hunk 33]
```diff
@@ -0,0 +1,78 @@
...
+
+# Server Events
+
+Events related to objects on a logical server.
+These classes are located in `net.fabricmc.fabric.api.event.lifecycle.v1`.
+
+## `ServerLifecycleEvents`
+
+Contains lifecycle events related to a Minecraft server.
+This includes events which indicate a when a server is starting up, is in a data pack reload or shutting down.
```
Comment (Technici4n) [inline fabric-lifecycle-events-v1/README.md]: a when a

**Review by Technici4n — APPROVED** [review]

**Review by Juuxel — APPROVED** [review]

**Review by YanisBft — APPROVED** [review]

**Review by modmuss50 — APPROVED** [review]

**Pull request merged**